File: src/cmcandy/java_language_answers/_0029.java
package cmcandy.java_language_answers;

/**
 * Created by CHEN on 2020/5/6.
 */
public class _0029 {
    static class Solution {
        public int divide(int dividend, int divisor) {
            if (divisor == -1 && dividend == -2147483648) return 2147483647;
            int symbol = 1;
            long divid = dividend;
            long divis = divisor;
            if (divis < 0) {
                symbol = -symbol;
                divis = -divis;
            }
            if (divid < 0) {
                symbol = -symbol;
                divid = -divid;
            } else if (divid == 0) return 0;
            if (divis == 1) return (int) (symbol > 0 ? divid : -divid);
            if (divis == divid) return symbol > 0 ? 1 : -1;
            if (divis>divid) return 0;
            long tmp = divis;
            int ans = 1;
            while (tmp+tmp < divid) {
                tmp =tmp<< 1;
                ans =ans<< 1;
            }
            ans = find_ans(tmp, ans, divid, tmp, ans);
            if (tmp > divid) ans--;
            return symbol > 0 ? ans : -ans;
        }

        public int find_ans(long tmp, int ans, long divid, long tmp_divis, int tmp_ans) {
            if (tmp + tmp_divis < divid) {
                tmp += tmp_divis;
                ans += tmp_ans;
            } else if (tmp + tmp_divis == divid){
                return ans + tmp_ans;
            }
            if (tmp_ans == 1) {
                return ans;
            }
            return find_ans(tmp, ans, divid, tmp_divis >> 1, tmp_ans >> 1);
        }

    }


    public static void main(String[] args) {
        Solution solution = new Solution( );
        System.out.println(solution.divide(10, 3));
        System.out.println(solution.divide(2, 2));
        System.out.println(solution.divide(-2147483648, -1));
        System.out.println(solution.divide(-2147483648, 2));
        System.out.println(solution.divide(1004958205, -2137325331));
        System.out.println(solution.divide(-1048958272, 336470323));

    }
}

File: src/cmcandy/Python_language_Answers/_0029.py
class Solution:
    def divide(self, dividend: int, divisor: int) -> int:
        '''
            1.先判断正负号
            2.将倍数转化成二进制
            3.注意int 和long之间的转化
        '''
        sign = 1 if (dividend > 0 and divisor > 0) or (dividend < 0
                                                       and divisor < 0) else -1
        a = abs(dividend)
        b = abs(divisor)

        time_bit_cnt = 0  #除法时，最高的bit位
        while a >= b:
            time_bit_cnt += 1
            b <<= 1

        time = 0
        while time_bit_cnt > 0:  #bit从左往右
            time_bit_cnt -= 1
            b >>= 1

            if a >= b:  #当前bit位，每次减1
                time += (1 << time_bit_cnt)
                a -= b

        if sign < 0:
            time = 0 - time
        return time if -(1 << 31) <= time <= (1 << 31) - 1 else (1 << 31) - 1